[1단계 - 지하철 노선도 미션] 콜린(이수연) 미션 제출합니다.

2SOOY opened the pull request:
안녕하세요 :) 콜린입니다 😄
어느덧 마지막 이네요! 마지막 미션도 잘 부탁드립니다!
항상 영양가 있는 리뷰 남겨주셔서 감사합니다 🙏   

[데모확인](https://doby-collin-react-subway.web.app/)

## 📌 구현 사항
+ 전반적으로 레벨 1의 지하철 미션과 동일합니다.
   + 로그인 / 회원가입 
   + 역 관리 - 추가 / 삭제
   + 노선 관리 - 추가 / 수정 / 삭제
   + 구간 관리 - 추가 / 삭제
+ 다만 이번 미션은 `백엔드와의 협업`이 추가되어 `API를 선택`할 수 있는 기능이 추가되었습니다.
   + 추가적인 기능을 위한 API보다는 서버마다 최대한 동일한 결과를 얻을 수 있도록 `동일한 형식의 에러 메시지`를 요청하였습니다 .

### ◻ 기술 스택
+ `언어` TypeScript
+ `상태관리` Redux Toolkit
+ `스타일` Emotion
+ `테스트` Storybook / RTL + Jest

## 📌 후기
1. **Redux Toolkit을 선택한 이유**
    + Redux를 제외한 다른 상태관리 도구는 고려를 하지 않았습니다.
    + 아직 Redux에 대한 학습이 충분히 이뤄지지 않았다고 판단되었기 때문입니다.
    + Redux Toolkit을 사용한 이유는 
        + 기존의 Redux의 `Action Type / Action / Reducer`등을 생성하는 반복적인 행위를 줄일 수 있어서 선택하였습니다.
        + Redux Toolkit를 사용한 경험이 없어 새로운 도구를 학습하는 차원에서 도입하였습니다.

2. **Saga를 도입하려다 포기한 이유**
    + Saga를 도입하려 했던 이유 => 여러 번의 `API Call`에 대한 간편한 처리
    + Saga에 대한 학습도 비용이므로 기간 내에 미션을 제출하지 못할 것 같다는 판단이 들어 나중에 도입하기로 합의

3. **지역 상태 vs 전역 상태**
    + `AccessToken`과 `로그인한 유저의 최소한의 정보` + `API 호스트 URL`를 `전역 상태`로 관리 
    + 그 외 지하철 역 / 노선 / 구간 등은 `지역 상태`로 관리
    + API 호스트 URL과 토큰, 유저정보는 다른 `모든 메뉴에서 공통적으로 사용`하기 때문에 전역으로 관리해야 한다고 판단
    + 나머지 역/노선/구간 해당하는 상태들은 `각각의 기능(메뉴)에 한정`되는 경우가 많아서 지역으로 관리

### ◻ 질문 & 요청
1. `타입스크립트`
이번에 `타입스크립트`를 처음 사용해봤는데요! 아직까지는 타입스크립트의 이점을 엄청나게 체감하진 못하고 있습니다ㅜ
타입스크립트의 이해도 많이 부족하기 때문인 것 같습니다. (안전하게 프로그래밍을 할 수 있다는 것은 새로 배웠습니다!)
**코드 전반적으로 타입스크립트를 적절하게 사용했는지 봐주셨으면 좋겠습니다 :)**

2. `컴포넌트의 역할`
레벨 2미션이 끝을 향해 가고 있지만 아직까지 하나의 컴포넌트에 얼마만큼의 역할을 부여할지 감이 잘 잡히지 않습니다.
대부분의 **상태는 최대한 상위 컴포넌트로 끌어올리고 props로 내려주는 방식을 택하는게 적절할까요?**
현재 프로젝트는 대부분의 상태를 각각의 Page 컴포넌트로 끌어올리게 설계하였습니다 !

      제가 생각하기에 컴포넌트는 각자가 독립적인 역할을 수행할 수 있을 수 있는 무언가라고 생각하는데, 하위 컴포넌트가 props를 많이 받는 순간 의존성이 발생하는 것이고 이는 독립적인 역할을 하는데 방해를 하는게 아닐까 하는 의문이 들었습니다.

      가령, `페이지` 컴포넌트 내에 `폼` 컴포넌트가 있는 경우 폼에 있는 `입력 값 + 입력값 핸들링`에 대한 것도 페이지 컴포넌트에 위임하는게 좋을지 아니면 폼 컴포넌트 독립적으로 상태와 상태 핸들러 등을 가지고 있는 것이 좋을지 발리스타님의 의견을 듣고 싶습니다 :) 

3. `FE단의 유효성 검사`
이번에 협업 미션을 진행하면서 프론트엔드에서 처리해야 할 유효성 검사에 대해서 궁금증이 들었습니다. 발리스타님은 유효성 검사에 대해 어떻게 처리하시는지 의견이 듣고 싶습니다 :) 
    **FE와 BE에서 모든 케이스에 대해서 전부 유효성 검사를 해줘야 할까요?**  

    예를 들어, `중복된 이름의 역을 추가하려고 시도`할 때 FE단에서 상태로 갖고 있는 모든 역을 조회해서 중복된 역이 있는지 판단하여 `요청 자체를 막는 것`과 우선은 BE에 `요청을 날리고` 그에 상응하는 적절한 `오류 메시지만 받아서 처리`하면 되는지 궁금합니다.(이번에는 적절한 에러 처리가 존재하는 API에 대해서는 후자의 방식을 택했습니다)
    
    
 **항상 감사드립니다 🙏🙏🙏**

woowapark committed:
Update README.md

Commit:
chore: ts CRA 구축 및 eslint, prettier 설정

Commit:
docs: Update README.md

Commit:
chore: 스토리북 환경 설정

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: Button 컴포넌트 구현

woowapark committed:
Update README.md

Commit:
feat: Input 컴포넌트 구현

Commit:
feat: Select 컴포넌트 구현

Commit:
feat: GlobalStyle 적용

Commit:
refactor: 디렉토리 구조 변경

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: Title 컴포넌트 구현

Commit:
feat: 공통 css 분리

Commit:
feat: Menu 컴포넌트 구현

Commit:
feat: RootContainer 컴포넌트 구현

Commit:
feat: Main 컴포넌트 구현

Commit:
feat: Header 컴포넌트 구현

Commit:
feat: ColorButton 컴포넌트 구현

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: Modal 컴포넌트 구현

Commit:
feat: ColorSelector 컴포넌트 구현

Commit:
feat: LoginForm 컴포넌트 UI 구현

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: ErrorMessage 컴포넌트 구현

Commit:
feat: SignUpForm 컴포넌트 구현

Commit:
feat: StationAddForm 컴포넌트 구현

Commit:
feat: LineAddForm 컴포넌트 구현

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: SectionAddForm 컴포넌트 구현

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: ListItem 컴포넌트 구현

Commit:
feat: LineEditForm 컴포넌트 구현

- LineAddForm 컴포넌트를 LineEditForm 컴포넌트와 병합

Commit:
refactor: API 명세에 맞는 타입 재정의

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: HOME 페이지 기능 구현 및 테스트 작성

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: SignUp 페이지 기능 구현 및 테스트 작성

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: Login 페이지 기능 구현 및 테스트 작성

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
refactor: 비동기 에러핸들링 및 signedUser 구조 변경

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: Menu 컴포넌트 기능 구현

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
refactor: 로그인 / 회원가입 기능 개선

- 에러 핸들링
- 커스텀훅으로 추출

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: Station 컴포넌트 기능 구현 및 테스트 작성

Commit:
feat: 백엔드 API 서버 선택기능 구현

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: Line 컴포넌트 노선 목록 조회 기능 구현

Commit:
feat: Line 컴포넌트 노선 추가/수정/삭제 기능 구현 및 테스트 작성

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: Section 컴포넌트 기능 구현

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
refactor: 기존 에러 메시지를 서버에서 수신한 에러 메시지로 대체

Commit:
feat: 로그아웃 기능 구현

Commit:
docs: Update README.md

Commit:
feat: redux-persist를 사용하여 새로고침 시에도 상태저장 기능 구현

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
feat: 역 이름 유효성 검사 기능 구현

Commit:
feat: 노선 이름 유효성 검사 기능 구현

Commit:
fix: 회원가입에서 에러 메시지가 처음에 렌더링되는 이슈 수정

Commit:
feat: 배경으로 Video 추가

Commit:
test: Section 컴포넌트 테스트 작성

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
refactor: 조건부 Routing 적용

Commit:
fix: accessToken이 유효하지 않은 경우 처리

Commit:
refactor: interface, enum 위치 변경

Commit:
refactor: 중복되는 style 추출 및 적용

Co-authored-by: Lee Soo Yeon <2SOOY@users.noreply.github.com>

Commit:
refactor: useServerAPI 커스텀훅 변경

Commit:
refactor: validation 로직 추출

Commit:
refactor: responseMessage 상수화

Commit:
refactor: api 요청 주소 상수화

Commit:
docs: Update README.md

2SOOY committed:
Merge branch 'step1' of https://github.com/zereight/react-subway-map into step1

2SOOY committed:
refactor: 코드 라인 순서 변경

- useServerAPI useEffect 클린업 제거

Vallista commented on .eslintrc:
```diff
@@ -0,0 +1,19 @@
+{
+  "plugins": ["prettier", "@typescript-eslint"],
```
tslint가 fade-out이라는 것을 알고 eslint로 적용하셨군요 굿굿

365kim replied:
혹시 프로그래밍에서 `fade-out` 은 무슨 뜻일까요...? `deprecated`랑 같은 의미로 쓰이는지, (인기가 없어지는 등의 이유로) 생태계에서 점점 사용되지 않는다는 의미인지 궁금합니다...! 👉👈

\+ tslint는 `deprecated` 라고 하니 `deprecated`의 의미인 것 같군용...! ㅎ_ㅎ

2SOOY replied:
사실 이 부분에 대해서는 전혀 인지하지 못하고 있었습니다,,ㅎㅎ
덕분에 `tslint`는 `depreacated`되었다는 사실을 알게되었습니다. 감사합니다 🙏

Vallista commented on .prettierrc:
```diff
@@ -0,0 +1,9 @@
+{
+    "semi": true,
```
semi를 true로 한 이유가 있을까요?

2SOOY replied:
코드를 작성할 때, 세미콜론을 붙이는 습관이 있기 때문에 `true`로 하였습니다.
또한 개인적으로 코드를 읽을 때도, 세미콜론이 있는 경우가 문(statement)마다 끝맺음을 명확하게 인지할 수 있기 때문에 붙이는 것을 선호합니다 :)

Vallista commented on .storybook/main.js:
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+  stories: ['../src/**/*.stories.mdx', '../src/**/*.stories.@(js|jsx|ts|tsx)'],
```
mdx로 사용하는 이유가 있나용?

2SOOY replied:
스토리북 기본 설정을 그대로 사용해서 큰 이유는 존재하지 않습니다,,ㅜ
현재는 `mdx`를 활용하지 않기 때문에 해당 부분은 제거하였습니다 :)

Vallista commented on README.md:
```diff
@@ -1,2 +1,92 @@
-⚛️ React 지하철 노선도 관리 페이지
+## 협업 요구사항
+
+- [x] step1에서 사용할 API의ㄴ 배포 일정, 이후 step2에서의 추가 개발을 위한 설계 논의, 그 외 필요한 사항들 있는 지 확인
```
오오오오~타!

2SOOY replied:
오타 수정하였습니다 😅

Vallista commented on package.json:
```diff
@@ -0,0 +1,75 @@
+{
+  "name": "react-subway-map",
+  "version": "1.0.0",
```
version이 1.0.0인 이유가 있나요?

2SOOY replied:
보통 App의 최초 배포를 진행할 때, '`1.0.0`버전부터 시작하지 않을까?' 하는 생각에서 버전을 변경했습니다.
하지만 여러 패키지를 확인한 결과 메이저 버전이 `0`부터 시작하는 패키지가 굉장히 많다는 것을 알 수 있었습니다 :)

덕분에 NPM version 각 자리의 의미 `major / minor / patch`에 대해 알 수 있었습니다. 감사합니다 🙏

Vallista commented on src/App.tsx:
```diff
@@ -0,0 +1,158 @@
+import { getSignedUserAsync } from './features/signedUserSlice';
+import { RootState, useAppDispatch } from './store';
+
+interface ConditionalRouteProps {
```
보통 I~를 붙이는게 interface의 양식인데, 이렇게 하면 Union Type Alias랑 햇갈리지 않을까요?

jbee37142 replied:
안녕하세요, @Vallista 님. 또 지나가다가 들렸습니다~

> 보통 I~를 붙이는게 interface의 양식인데

동의하지 않습니다!

**일관성을 파괴하는 네이밍 컨벤션**
여기에서 I는 인터페이스 임을 드러내는 것이며 이것은 네이밍에서부터 타입을 드러내는 헝가리식 표기법입니다.
하나의 프로젝트 (또는 어떤 다른 기준)에서 snake_case를 사용한다던가, camelCase를 사용한다던가 이런 표기법은 일관성이 가장 중요합니다.

다른 변수나 함수 네이밍에는 헝가리식 표기법을 적용하지 않다가 인터페이스에만 헝가리식 표기법을 적용하는 것은 잘못된 것입니다.

**컨벤션의 목적과 사용되는 언어가 근본적으로 맞지 않는 문제**
더군다나 자바스크립트와 같은 언어에서 변수 (또는 함수)에 자료형을 드러내기 위해 사용하던 헝가리식 표기법을 구조적 타이핑을 기반으로 하는 TypeScript 에 적용하는 것은 더욱 말이 안 됩니다.

cc. https://github.com/microsoft/TypeScript-Handbook/issues/121#issuecomment-628193866

Vallista replied:
@JaeYeopHan 오 이 답글을 지금 보았네요!

되게 오랜만에 듣는 헝가리식 표기법이네요! ㅋㅋㅋ 말씀하시는 부분에 대해서 동의합니다 👍 
그렇다면, 두 개에 대한 구분을 위해 재엽님은 어떤 prefix 같은 양식을 주시는 지 궁금합니다~

해당 부분에 대해서 구분이 필요한 이유는, 사용될 때 interface의 특성이나 type 적인 특성을 구별하기 위함인데, 요 것에 대해서는 어떻게 생각하시나요?

2SOOY replied:
우선적으로 발리스타님처럼  
> interface의 특성이나 type 적인 특성을 구별하기 위함

일부러 인터페이스에 `I prefix`를 활용했던 것 같습니다. 이야기 해주신 것처럼 `Union Type Alias`랑 헷갈릴 수도 있을 것 같습니다 !!

`I prefix`를 많이 사용해서 이번 미션은 그대로 진행하도록 하겠습니다 😢 

추가로 타입스크립트 [코딩 가이드라인](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines)에 따르면 "인터페이스에 `I prefix` 사용하지 않기"가 있어 다음 프로젝트 진행시 `I prefix`는 지양하도록 하겠습니다 :)
![image](https://user-images.githubusercontent.com/46412689/124348877-7afa6e80-dc27-11eb-8ee0-32800d5b5d1e.png)

Vallista commented on src/App.tsx:
```diff
@@ -0,0 +1,158 @@
+interface ConditionalRouteProps {
+  exact?: boolean;
+  path: string;
+  Component: React.ComponentType;
```
React.ReactElement 요런게 더 낫지 않을지..? 아니면 React.ReactNode

2SOOY replied:
e9f58c3

`ConditionalRoute` 컴포넌트에서 `Component` prop 필요 없이 구현 가능해서 해당 prop을 제거하였습니다 :)

Vallista commented on src/App.tsx:
```diff
@@ -0,0 +1,158 @@
+  condition: boolean;
+}
+
+const ConditionalRoute = ({ exact, path, Component, condition }: ConditionalRouteProps) => {
```
요거 뭔가 하나의 파일 (ex: route.ts) 로 빼서 관리할 수 있어보여요~

2SOOY replied:
5633c47

`App` 컴포넌트가 너무 많은 정보를 갖고 있어서 고민이었는데, 덕분에 Route 부분 분리하여 App 컴포넌트를 간단하게 표현할 수 있었습니다.
([🔽중복된 부분 map만들기](https://github.com/woowacourse/react-subway-map/pull/9#discussion_r642029256)에 대한 리뷰도 같이 적용하였습니다)

Vallista commented on src/App.tsx:
```diff
@@ -0,0 +1,158 @@
+        <Switch>
+          <Route exact path={ROUTE.HOME} component={Home} />
+
+          <ConditionalRoute
```
중복되는 정보들이 많은데 배열로 만들어서 map으로 맹그는건 어떨까요?

2SOOY replied:
`map`으로 관리하니 깔끔하게 관리할 수 있네요 👍👍👍 감사합니다 🙏  
(+ 추가로 메뉴 관련 부분도 `Links`컴포넌트로 관리하도록 변경하였습니다. 42fc622)

Vallista commented on src/App.tsx:
```diff
@@ -0,0 +1,158 @@
+      <Button
+        type="button"
+        buttonTheme="menu"
+        onClick={() => history.push({ pathname: ROUTE.STATION })}
```
goToHistory 요런 함수 만들어서, Route.LINE등을 쉽게 갈 수 있도록 함수 만드는게 어떨까요? 중복로직이 많네용

Vallista replied:
그리고 enum으로 정의하는게 더 나을까요? type alias로 정의하는게 나을까요?
저는 요 부분은 enum도 좋지만 type alias로 해도 별 상관없겠다는 생각이 드네용.
괜히 ROUTE도 포함해야하니..

2SOOY replied:
42fc622

페이지 이동 부분에 대해 `goPage`함수로 분리하여 관리하도록 변경하였습니다 :)

2SOOY replied:
여러 자료들을 봤지만 각각을 사용해야 할 시점에 대해 명확히 인지하진 못한 상태입니다. 😢 

다만 `type alias`로 관리하게 된다면 자동완성 없이 특정 값을 직접 입력 해야 하는 번거로움이 있을 것 같습니다. (IDE를 활용하면 직접 입력하지 않아도 되겠지만요,,,ㅎㅎ) 
반면, `enum` 혹은 `객체 as const`를 활용한다면 자동완성을 이용할 수 있는 이점이 존재할 것 같습니다.

Vallista commented on src/components/atoms/Button/Button.styles.ts:
```diff
@@ -0,0 +1,37 @@
+import { css } from '@emotion/react';
```
따로 분리한 것 좋네요!

Vallista commented on src/components/atoms/Button/Button.styles.ts:
```diff
@@ -0,0 +1,37 @@
+
+const buttonThemeTable = {
+  default: css`
+    background-color: #4ecfe0;
```
컬러값은 constant로 만들어서 한 군데에서 관리해주세요~

Vallista replied:
관리를 한 군데에서 안하면 추후 유지보수때 힘들어집니다!

2SOOY replied:
b446ed1

넵! `컬러`값에 대해 상수화하여 한 군데에서 관리하도록 진행하였습니다 :)

Vallista commented on src/components/atoms/Button/Button.tsx:
```diff
@@ -0,0 +1,26 @@
+import { Container } from './Button.styles';
+
+export type ButtonTheme = 'default' | 'edit' | 'menu';
```
export될 필요가 있나요?

2SOOY replied:
7753772

해당 `ButtonTheme 타입`은 `Button.styles.ts`로 옮긴 상태이고, `Button`컴포넌트의 props로 `ButtonTheme 타입`이 필요해서 export 키워드 자체는 유지하였습니다 :)

```ts
// Button.styles.ts
export type ButtonTheme = 'default' | 'edit' | 'menu';
```

```tsx
// Button.tsx

interface ButtonProps {
  buttonTheme?: ButtonTheme
}

```

Vallista commented on src/components/atoms/ColorButton/ColorButton.tsx:
```diff
@@ -0,0 +1,12 @@
+  onClick: React.MouseEventHandler<HTMLButtonElement>;
+}
+
+const ColorButton = ({ bgColor, onClick }: ColorButtonProps) => (
```
Button과 무엇이 다르기 때문에 요 것이 따로 분리 되었을까요..? 프리미티브 컴포넌트(Button)를 사용하면 될 듯 한데요~

2SOOY replied:
7753772

살펴보니 사실상 `Button`과 다를게 없군요 ! `ColorButton` 컴포넌트에 대한 하나의 버튼 테마(colorPick)를 추가하여 관리하도록 변경하였습니다 :)

```ts
export type ButtonTheme = 'default' | 'edit' | 'menu' | 'colorPick';

```

Vallista commented on src/components/atoms/Header/Header.tsx:
```diff
@@ -0,0 +1,12 @@
+}
+
+const Header = ({ hasExtra = false, children }: HeaderProps) => {
+  return <Container hasExtra={hasExtra}>{children}</Container>;
```
```tsx
const Header = ({ children, ...otherProps }: HeaderProps) => {
return <Container {...otherProps}>{children}</Container>
}
```

로 표현할 수도 있습니다. 호홍~ 바이패스되면 요렇게 하는게 더 낫지 않을까요?

2SOOY replied:
78b4dd3

미션 진행하면서 단순히 props를 전달하는 부분에서 `...props`를 잘 활용하지 못했었네요
수정완료하였습니다 :)

Vallista commented on src/components/atoms/Input/Input.styles.ts:
```diff
@@ -0,0 +1,11 @@
+import styled from '@emotion/styled';
+
+const Container = styled.input`
```
Container로 명명한 이유가 있을까요..? 흠.. 맞지않는 이름 같습니다.

"container는 데이터의 모음을 격리하는 공간" 으로 많이 쓰셔요. 근데 여기서 쓰임새는 style을 적용하는 목적으로만 쓰이는데요!
차라리 ElementWrapper 라던가. 좀 더 스타일에 한정된 이름으로 만드는게 좋겠습니다.

2SOOY replied:
습관적으로 컴포넌트 파일 내에서 가장 상위에 해당하는 부분을 `Container`라고 명명했는데,
확실히 맞지 않은 이름이라고 생각합니다 ! 따라서 `StyledInput`처럼 앞에 `Styled`라는 `prefix`를 활용하는 방식으로 변경하였습니다 :)

Vallista commented on src/components/atoms/Input/Input.tsx:
```diff
@@ -0,0 +1,33 @@
+import { Container } from './Input.styles';
+
+export interface InputProps {
+  placeholder: string;
```
bypass만 해줄 것이라면, `InputProps extends React.CSSProperties` 만으로도 충분할 수도 있겠네요. 아니면 Input Properties만 갖고있는 친구가 있나..?

2SOOY replied:
fd6ed5d

전반적으로 기존 `HTMLElement`의 `HTMLAttributes` 속성을 확장하지 않고 사용했었네요,,ㅜ
따라서 
`interface InputProps extends InputHTMLAttributes<HTMLInputElement>`와 같은 방식으로 인터페이스를 지정하고
꼭, props로 받아야만 하는 html 속성값만 받도록 변경하였습니다 :)

Vallista commented on src/components/atoms/Input/Input.tsx:
```diff
@@ -0,0 +1,33 @@
+}: InputProps) => (
+  <Container
+    type={type}
+    placeholder={placeholder}
```
placholder나 type도 ...props로 대응되지 않나요? 따로 빠져있는 이유가 무엇인가요?
text는 default type을 선언하기 위함으로 이해가 되는데, placeholder는요!?

2SOOY replied:
이 부분 위의 PR과 함께 적용하였습니다 :) 덕분에 `...props`에 대해 명확하게 알 수 있었습니다 🙏

Vallista commented on src/components/atoms/Menu/Menu.stories.tsx:
```diff
@@ -0,0 +1,31 @@
+Basic.args = {
+  children: (
+    <div>
+      {['역 관리', '노선 관리', '구간 관리', '경로 조회'].map(text => (
```
`['역 관리', '노선 관리', '구간 관리', '경로 조회']`는 변경되지 않으므로, args 밖으로 빼내는게 좋지 않을까요? args 데이터가 변경되면 배열도 계속 새로 생성해서 넣겠군요..?

2SOOY replied:
d86e809

해당 부분에 대해서는 전혀 인지하지 못하고 있었는데, 짚어주셔서 감사합니다 🙏
해당 배열은 외부로 분리하였습니다 :)

Vallista commented on src/components/molecules/HostSelect/HostSelect.styles.ts:
```diff
@@ -0,0 +1,20 @@
+const Container = styled.div`
+  position: fixed;
+  top: 0;
+  right: 5%;
```
5%인 이유가 있나욤? 다른친구들은 다 rem인데..

Vallista replied:
5% 크기가 보장이 안될터인데..

2SOOY replied:
afa294a

화면 크기에 관계 없이 5% 만큼의 우측에 위치하도록 의도하였습니다.
다만 정말 그럴 필요가 있을 지에 대한 의문이 드네요! 추가로 발리스타님이 남겨주신 `크기 보장`에 대한 문제가 존재하군요.
따라서 `rem`으로 변경하였습니다.

Vallista commented on src/components/molecules/LineEditForm/LineEditForm.tsx:
```diff
@@ -0,0 +1,82 @@
+    })) || [];
+
+  return (
+    <Container onSubmit={onSubmitLineInfo}>
```
Container라고 하니 이 친구가 Form인지 뭔지 모르겠네요 ㅠ

2SOOY replied:
해당 부분도 `StyledForm`으로 이름 변경하였습니다 :)

Vallista commented on src/components/molecules/SectionAddForm/SectionAddForm.tsx:
```diff
@@ -0,0 +1,82 @@
+  lineList: ILineRes[];
+  onSubmitSectionInfo: React.FormEventHandler<HTMLFormElement>;
+
+  onChangeLine: React.ChangeEventHandler<HTMLSelectElement>;
```
Form을 처리하기위해 이러한 데이터를 다 받기보다 Context API의 사용을 검토해봣으면 합니다~

2SOOY replied:
9eeef01

`Context API`를 활용해보았습니다. `Context API`를 활용하면서 확실히 모든 state를 전부 선언할 필요가 없어져서 페이지 컴포넌트가 간소화 되는 것을 겪었습니다. 다만, `Form`처리와 관련하여 어디까지 `Provider`로 제공할지 (onSubmit, onChange 등등) 고민이 필요할 것 같습니다.

Vallista commented on src/components/molecules/SignUpForm/SignUpForm.tsx:
```diff
@@ -0,0 +1,87 @@
+import { Button, ErrorMessage, Input } from '../../atoms';
+
+export interface SignUpFormProps {
+  onChangeAge: React.ChangeEventHandler<HTMLInputElement>;
```
Form을 제어하기위해 이렇게 다 받기보다 Context API를 사용하는게 좋아보이고, 이 부분이 성능적 우려가 된다면 useState를 Object 형태로 여러개를 만들어서 관리를 한 번에 하는게 더 낫지 않은 선택지가 아닐지 고민해보면 좋겠습니다.

2SOOY replied:
e23c489

해당 `SignUp`페이지에 대해서도 `Context API`를 활용하는 방식으로 변경해보았습니다 ! 
발리스타님이 주신 참고 링크와 `Context API` + `Form`과 관련된 자료를 보면서 다른 방식으로도 생각이 필요할 것 같습니다 :)

Vallista commented on src/components/pages/Home/Home.test.tsx:
```diff
@@ -0,0 +1,43 @@
+    expect(container).toHaveTextContent(GUIDE_MESSAGE.BEFORE_LOGIN);
+  });
+
+  it('render before-login guide message', () => {
```
after-login이 아닐까요?

2SOOY replied:
그렇네요,, 꼼꼼하게 봐주셔서 감사합니다 🙏 
사소한 것도 다시 살펴보도록 하겠습니다 :)

Vallista commented on src/components/pages/Line/Line.tsx:
```diff
@@ -0,0 +1,208 @@
+import { LineItemWithCircle, ListItemContainer } from './Line.styles';
+
+const isValidLineName = (lineName: string) => {
+  return /^[가-힣0-9]{2,10}$/.test(lineName);
```
요런 validation Regex를 따로 때서 관리하는 건 어떨까요? 흔히 사용될 수 있어보이네요!

2SOOY replied:
b446ed1

넵! `validation`관련 로직 유틸로 분리하였습니다 :)

Vallista commented on src/components/pages/Line/Line.tsx:
```diff
@@ -0,0 +1,208 @@
+    onChange: onChangeDistance,
+    setValue: setDistance,
+  } = useChangeEvent('1');
+  const {
```
너무 상태를 파편화 한 게 아닌지...

2SOOY replied:
f7f1c93

이 부분에 대해서도 `Context API`를 활용해보았습니다 :) 감사합니다 🙏

Vallista commented on src/components/pages/Login/Login.test.tsx:
```diff
@@ -0,0 +1,82 @@
+    (useDispatch as jest.Mock).mockImplementation(() => jest.fn());
+  });
+
+  it('로그인 페이지에서 유효한 이메일/비밀번호를 입력 후 로그인 버튼을 눌렀을 때, 홈페이지로 이동한다.', () => {
```
이동하지 못하는 경우에 대한 테스트 케이스는 없나요?

Vallista commented on src/components/pages/Login/Login.test.tsx:
```diff
@@ -0,0 +1,82 @@
+    expect(window.location.pathname).toEqual(ROUTE.SIGNUP);
+  });
+
+  it('로그인 페이지에서 회원가입 버튼을 눌렀을때, 회원가입 페이지로 이동한다.', () => {
```
테스트는 정상적으로 잘 작동되는 경우보다, 정상적으로 작동되지 못하는 경우에 대한 테스트가 더 중요합니다.

2SOOY replied:
다음 프로젝트에서 테스트 코드 작성 시 비정상적인 부분 우선으로 작성하도록 하겠습니다 !

Vallista reviewed (changes requested):
안녕하세요 콜린!

대망의 마지막 지하철 노선도 미션이군요~ 이 미션도 함께 잘 부탁드립니다!
코드가 엄청 많아서 깜짝놀랐습니다. 코드리뷰가 점점 어려워지는 걸 느끼네요! ㅋㅋ..
그 만큼 콜린님이 정성을 담아 코드를 짜셨음을 알기에 마지막 미션인만큼 더 빡세게 코드리뷰를 진행했슴니다!

# 전체적인 피드백

1. 전체적으로 조금씩 공통화로 나눌 수 있는 부분이 있음에도 제대로 나눠지지가 않을 것 같습니다. 예를들어 유틸리티 영역으로 뺄 수 있는 컬러라던지, regex라던지, 개발 일정이 촉박하여 공통화를 못한 부분들이 있어요. 그래서 이러한 부분은 피드백을 살짝쿵 남겨두었으니 확인해보시고, 수정해주시면 감사하겠습니다~

2. Type Alias와 Enum을 사용하는 패턴에서 적절한 곳에 사용하고 있는지, TypeScript 내에서 어떤 상황에 Enum과 Type Alias를 사용하면 좋은지 등을 다시 한번 고려해보면 좋겠습니다 👍 

3. Redux Toolkit을 사용하셨는데, Redux Toolkit은 기존의 Redux의 반복적으로 작성하는 행위에 대해서 해결하기 위해 어떻게 접근하고 있고, 콜린님이 사용하시면서 어떤 생각이 드셨나요? Redux Toolkit의 가장 중요한 개념은 무엇일까요? 사실, "Redux Toolkit를 사용한 경험이 없어 새로운 도구를 학습하는 차원에서 도입하였습니다." 라고 적어주셨는데, Redux Toolkit은 "새로운 도구를 학습하는 차원보다는, 기존의 어떤 형태에 반해, 어떠한 형태가 좋기 때문에 이러한 형태를 제공하는 부분이 적합하다 생각하여 적용했습니다." 로 이해를 하고 사용을 하는게 더 좋았을 것 같습니다. 요 부분은 Request Changes로 던질테니, 코맨트로 달아주시면 감사하겠습니당~ (바쁘지만 중요한 부분이고, 리덕스의 사용형태에 대해 공부를 하는 부분이기도 하니 꼭 생각해보시길 추천드려요!)

4. Context API는 굳이, 전역이 아닌 Context의 뜻인 "맥락", "문맥" 등의 상태를 저장하는데 좋은 요소일 수 있습니다. Form을 많이 쓰시는 만큼, 다양한 Form 단위로 상태와 상태를 변경하는 함수를 한 쌍으로 해서 몇 개씩 Props로 받고 있는데요! 과연 이게 맞을까요..? 다양한 상태 변화에 따라 의존도가 높아지고, 내부에서 처리해야할 상황이 많아지니, 요 부분을 효율적으로 하나의 Pipeline을 뚫어 사용하는 부분으로 Context API를 사용해서 해소할 수 있겠습니다. Form 하나당 Context API를 하나씩 만들어 수정하는 부분도 한번 고려해보세요~ 이러한 형태로 다양한 디자인 시스템 혹은 라이브러리들이 문제를 해결하고 있습니다. (예시: [Adobe React Design System - React Spectrum의 Form](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/form/src/Form.tsx))

5. custom hooks를 적극적으로 사용하려고 한 부분이 보입니다! 요 부분은 아주 칭찬쓰!! 다만, 지금과 같이 각잡고 공통화된 부분에만 쓰지마시고, useState하나만 써도 막 커스텀 훅으로 빼보세요. 수많은 hooks들이 나올 것인데, 한 훅들 10개이상 생성했을 때 쯤 내부 로직의 공통적인 부분을 한번 봐보세요. 그러면 공통된 부분이 보일꺼에요. 그런 것들을 하나씩 공통화 시켜보면서 비즈니스 로직도 넣어보고 여러부분을 가감없이 사용해보세요. custom hook 쓴다고 성능 나빠지는 것도 아니고, 더 막 써보시길 권합니다. 너무 막쓰면,, 리뷰어가 알아서 피드백을 하지 않을까요,,,? (무섭)

6. 테스트 케이스가 올바른 상황에 대해서만 존재하는데, 사실 올바른 상황보다 올바르지 않은 상황에 빛을 발하는게 테스트 코드입니다. 역으로 말하면 테스트 코드가 제 역할을 못하고 눈가리고 아웅만 하는 느낌이네요 ㅠㅠ. 정상적인 상황도 좋지만, 오히려 시간이 없다면 정상적인 상황이 아닌 "정상적으로 실패하는 상황" 혹은 "정상적으로 실패하지 못한 엣지 케이스인 상황"에 대해 테스트 코드 짜는게 더 낫겠습니다. TDD를 하는게 아니라, 테스트 코드를 이후에 넣는 상황이니까요.

7. Emotion으로 만든 스타일 엘레먼트를 container라고 표현했는데, 맞지않은 표현으로 보입니다. Container가 의미하는게 무엇이고, Container & Presentational Pattern은 왜 Container인지..? 를 한 번 확인해보면 좋겠어요.

8. Bypass 되는 부분이 많은데,,, Emotion을 쓰신다면 CSS Props 타입을 활용하여 Bypass 시키는게 어떠실 지..?

9. 나머지 상세한 부분은 코드 라인에 코맨트 달았습니다 👍 

# 질문 답변

## 1번째 질문

```
타입스크립트
이번에 타입스크립트를 처음 사용해봤는데요! 아직까지는 타입스크립트의 이점을 엄청나게 체감하진 못하고 있습니다ㅜ
타입스크립트의 이해도 많이 부족하기 때문인 것 같습니다. (안전하게 프로그래밍을 할 수 있다는 것은 새로 배웠습니다!)
코드 전반적으로 타입스크립트를 적절하게 사용했는지 봐주셨으면 좋겠습니다 :)
```

일단 PropTypes등으로 귀찮게 따로 코딩 안해도 타입 체킹이 되니 어느정도 안정성이 체감은 되실꺼구요! 타입에 맞지 않거나 하는 부분을 빌드타입에서 타입스크립트가 잡아주니 요 부분은 알게모르게 이점을 보고 계실꺼에요.

다만, Type Alias가 유효한 상황과, interface가 유효한 상황 등등은 코드 라인에 피드백한 것 처럼 고민해보시길 권합니다~

그 외는 아직 문법적으로 이점을 느끼기에 사용하신 부분이 없어서 딱히 피드백 드릴 요소가 없네요.. variadic tuple을 generic으로 묶어쓴다거나 뭐 그러한 레벨까지 아닌 이상은...

## 2번째 질문

```
컴포넌트의 역할
레벨 2미션이 끝을 향해 가고 있지만 아직까지 하나의 컴포넌트에 얼마만큼의 역할을 부여할지 감이 잘 잡히지 않습니다.
대부분의 상태는 최대한 상위 컴포넌트로 끌어올리고 props로 내려주는 방식을 택하는게 적절할까요?
현재 프로젝트는 대부분의 상태를 각각의 Page 컴포넌트로 끌어올리게 설계하였습니다 !
```

요 부분은 위에 피드백들 드렸습니다! 4번을 확인해주세요! 만약 컨텍스트 API의 활용이 성능적/문법적 으로 어렵다면 적어도 상태를 도메인 단위로 useState를 하나만 만들어서 전달하던지.. 그러한 방법으로 props로 던져주는 상태를 줄였으면 합니다. 코드가 너무 보기 어려워용~~ ㅋㅋ ㅠㅠ

## 3번째 질문

```
제가 생각하기에 컴포넌트는 각자가 독립적인 역할을 수행할 수 있을 수 있는 무언가라고 생각하는데, 하위 컴포넌트가 props를 많이 받는 순간 의존성이 발생하는 것이고 이는 독립적인 역할을 하는데 방해를 하는게 아닐까 하는 의문이 들었습니다.

가령, 페이지 컴포넌트 내에 폼 컴포넌트가 있는 경우 폼에 있는 입력 값 + 입력값 핸들링에 대한 것도 페이지 컴포넌트에 위임하는게 좋을지 아니면 폼 컴포넌트 독립적으로 상태와 상태 핸들러 등을 가지고 있는 것이 좋을지 발리스타님의 의견을 듣고 싶습니다 :)
```

리액트에서 제공하는 Props는 IoC의 DI를 따르고 있는 패턴입니다. 요 부분을 통해서 의존성을 분리시킬수도 있지만, 지금은 상태가 bypass됨으로써 비즈니스상으로는 강력한 결합이 되었는데요~ 이러할 때 Context API (4번 참조)를 이용해서 IoC의 Service Locator로 사용하는것도 좋겠습니다.. (요것도 4번 참고)

```
FE단의 유효성 검사
이번에 협업 미션을 진행하면서 프론트엔드에서 처리해야 할 유효성 검사에 대해서 궁금증이 들었습니다. 발리스타님은 유효성 검사에 대해 어떻게 처리하시는지 의견이 듣고 싶습니다 :)
FE와 BE에서 모든 케이스에 대해서 전부 유효성 검사를 해줘야 할까요?

예를 들어, 중복된 이름의 역을 추가하려고 시도할 때 FE단에서 상태로 갖고 있는 모든 역을 조회해서 중복된 역이 있는지 판단하여 요청 자체를 막는 것과 우선은 BE에 요청을 날리고 그에 상응하는 적절한 오류 메시지만 받아서 처리하면 되는지 궁금합니다.(이번에는 적절한 에러 처리가 존재하는 API에 대해서는 후자의 방식을 택했습니다)
```

당연히 FE, BE다 예외처리 해야하구요. FE에서 예외처리하는 이유는 클라이언트의 에러방지 - 사용자의 사용성 측면, BE에서 예외처리하는 이유는 데이터의 손실 및 시스템적 오류의 방지 (ex: -2000원이 결제금액인데 결제가 되었다고 하면 정산상으로 2000원을 돌려줘야 하는 것이니.. 무조껀 시스템상으로 안넘어가게 예외를 해야겠죠?) 둘 다 필요합니다.

다만 위의 예시의 관해서는 다양한 이유가 있을텐데요. 서버상에서 모든 역을 조회해서 중복된 역을 찾는 코드가 성능적으로 문제가 있다면 클라이언트에서 막을 것이고, 해당 로직이 문제가 없다면 서버에서 막아도 되겠네요. 이러한 로직이 어떤 사이드에 있는지에 대한 문제는 유동적으로 판단하시면 되겠습니다. 그런데 클라이언트 단에서 그리 처리하기 어려운 건 아니니 클라쪽에서 해도 되지 않을까요? 다만 혹시라도 중복된 역이 클라에서는 없었는데 서버에서 생긴 경우 (패칭이 제대로 안된경우?) 가 있으니 서버의 성능적 이슈가 없다면 서버에 있는게 더 좋겠네요.

2SOOY committed:
refactor: 컬러 상수 및 유틸성 함수 위치 이동

- isValidLineName
- 버튼 HEX 값

2SOOY committed:
feat: ConditionalRoute 컴포넌트 분리

- 기존 직접 작성한 Props 대신 RouteProps를 확장하여 인터페이스 작성
- condition에 따라 Route or Redirect 컴포넌트 렌더

2SOOY committed:
feat: 라우팅 관련 컴포넌트 Routes로 분리 및 배열을 활용한 라우팅

2SOOY committed:
fix: 서버 호스팅 URL 재설정

2SOOY committed:
refactor: 메뉴 버튼 모음 Links 컴포넌트로 분리

2SOOY committed:
refactor: 폴더 이름 변경 features -> modules

2SOOY committed:
refactor: Header 컴포넌트 props bypass하는 것 ...props로 변경

- props 설정시 기존 HTMLElement의 attributes 확장하면서 설계하기

2SOOY committed:
refactor: Input 기본 속성 ...props로 대체 및 name 속성 추가

- Input 컴포넌트 사용부 name prop 추가
- Input 컴포넌트 내 기본적인 Input 속성 제거

2SOOY committed:
refactor: ColorButton 제거 및 Button에 ColorPick 테마 추가

2SOOY committed:
refactor: FlexCenterBox에 shared style css 적용

2SOOY committed:
feat: Form을 다루기 위한 FormContext 생성

- context API 활용
- FormProvider를 통해 form state / onChange / onSubmit을 제공
- onChange를 위한 reducer 함수 추가

2SOOY committed:
feat: 회원가입 관련 FormProvider 적용

- Input 컴포넌트 FormProvider의 onChnage 사용
- 회원가입 페이지 state 선언 로직 제거
- validation 로직 SignUpForm 컴포넌트로 이동

2SOOY committed:
refactor: 로그인 페이지 FormContext 적용

2SOOY committed:
feat: FormContext 컴포넌트 reset액션 추가

2SOOY committed:
refactor: 역 관리 페이지 FormContext 적용

- useStation 커스텀 훅 추가

2SOOY committed:
feat: FormContext를 사용하는 별도의 ContextSelect 컴포넌트 생성

- Container 이름 -> StyledSelect로 변경

2SOOY committed:
refactor: Line 페이지 FormContext 적용

- 노선추가/수정 폼 분리
- ColorSelector 컴포넌트 FormProvider의 dispatch 사용
- FormProvider에서 submitFunc와 dispatch도 전달하도록 수정
- 노선과 관련된 useLine 커스텀훅 생성

2SOOY committed:
refactor: Section 페이지 FormContext 적용

2SOOY committed:
refactor: 오타 수정 및 불필요한 부분 제거

- README 오타 수정
- Home화면 테스트 이름 수정
- storybook 사용하지 않는 확장자 제거
- Menu.stories.tsx: 메뉴 이름 배열 외부에 선언

2SOOY committed:
refactor: Header 컴포넌트 스타일 관련 Props 위치 변경

2SOOY committed:
fix: HostSelect 컴포넌트 % 속성값 고정값으로 변경

2SOOY committed:
fix: 액션 함수를 활용하기 위한 export ActionType

2SOOY committed:
refactor: 역 관리 페이지 FormContext 사용 로직 다른 페이지 구조와 동일하게 변경

2SOOY committed:
feat: 메인 Video 컴포넌트 활성화
